pennylane/templates/embeddings.py

```diff
@@ -17,10 +17,85 @@
 from pennylane.ops import RX, RY, RZ, BasisState, Squeezing, Displacement, QubitStateVector
-from pennylane.variable import Variable
-import numpy as np
+from pennylane.templates.utils import (_check_shape, _check_no_variable, _check_wires,
+                                       _check_hyperp_is_in_options, _check_type)
+
+
+def AmplitudeEmbedding(features, wires, pad=None, normalize=False):
+    r"""Encodes :math:`2^n` features into the amplitude vector of :math:`n` qubits.
+
+    If the total number of features to embed is less than the :math:`2^n` available amplitudes,
+    non-informative constants (zeros) can be padded to ``features``. To enable this, the argument
+    ``pad`` should be set to ``True``.
+
+    The L2-norm of ``features`` must be one. By default, AmplitudeEmbedding expects a normalized
+    feature vector. The argument ``normalize`` can be set to ``True`` to automatically normalize it.
+
+    .. note::
+
+        AmplitudeEmbedding uses PennyLane's :class:`~pennylane.ops.QubitStateVector` and only works in conjunction with
+        devices that support this operation.
+
+    Args:
+        features (array): input array of shape ``(:math:`2^n`,)``
+        wires (Sequence[int] or int): int or sequence of qubit indices that the template acts on
+
+    Keyword Args:
+        pad (float or complex): if not None, the input is padded with this constant to size :math:`2^n`
+        normalize (Boolean): controls the activation of automatic normalization
+
+    Raises:
+        QuantumFunctionError if inputs do not have the correct format.
+    """
+
+    #############
+    # Input checks
+    _check_no_variable([pad, normalize], ['pad', 'normalize'])
+
+    msg = "At this stage, the feature input of AmplitudeEncoding cannot be trained. " \
+           "It has to be passed to the qnode via a positional argument."
+    _check_no_variable([features], ['features'], msg=msg)
+
+    wires, n_wires = _check_wires(wires)
+
+    n_ampl = 2**n_wires
+    if pad is None:
+        msg = "AmplitudeEmbedding must get a feature vector of size 2**len(wires), which is {}. Use 'pad' " \
+               "argument for automated padding.".format(n_ampl)
+        shp = _check_shape(features, (n_ampl,), msg=msg)
+    else:
+        msg = "AmplitudeEmbedding must get a feature vector of at least size 2**len(wires) = {}.".format(n_ampl)
+        shp = _check_shape(features, (n_ampl,), msg=msg, bound='max')
+
+    _check_type(pad, [float, complex, type(None)])
+    _check_type(normalize, [bool])
+    ###############
+
+    # Pad
+    n_feats = shp[0]
+    if pad is not None and n_ampl > n_feats:
+        features = np.pad(features, (0, n_ampl-n_feats), mode='constant', constant_values=pad)
+
+    # Normalize
+    norm = 0
+    for f in features:
+        if isinstance(f, Variable):
+            norm += np.conj(f.val) * f.val
+        else:
+            norm += np.conj(f) * f
+    norm = np.real(norm)
+
+    if not np.isclose(norm, 1.0, atol=1e-3):
```

Comment:
magic numbers should go at top of file. Also, we usually follow the pattern to also set `rol=0` explicitly